# -*- coding: utf-8 -*-
############################################################
#
# Autogenerated by the KBase type compiler -
# any changes made here will be overwritten
#
############################################################

from __future__ import print_function
# the following is a hack to get the baseclient to import whether we're in a
# package or not. This makes pep8 unhappy hence the annotations.
try:
    # baseclient and this client are in a package
    from .baseclient import BaseClient as _BaseClient  # @UnusedImport
except ImportError:
    # no they aren't
    from baseclient import BaseClient as _BaseClient  # @Reimport


class fba_tools(object):

    def __init__(
            self, url=None, timeout=30 * 60, user_id=None,
            password=None, token=None, ignore_authrc=False,
            trust_all_ssl_certificates=False,
            auth_svc='https://ci.kbase.us/services/auth/api/legacy/KBase/Sessions/Login',
            service_ver='release',
            async_job_check_time_ms=100, async_job_check_time_scale_percent=150, 
            async_job_check_max_time_ms=300000):
        if url is None:
            raise ValueError('A url is required')
        self._service_ver = service_ver
        self._client = _BaseClient(
            url, timeout=timeout, user_id=user_id, password=password,
            token=token, ignore_authrc=ignore_authrc,
            trust_all_ssl_certificates=trust_all_ssl_certificates,
            auth_svc=auth_svc,
            async_job_check_time_ms=async_job_check_time_ms,
            async_job_check_time_scale_percent=async_job_check_time_scale_percent,
            async_job_check_max_time_ms=async_job_check_max_time_ms)

    def build_metabolic_model(self, params, context=None):
        """
        Build a genome-scale metabolic model based on annotations in an input genome typed object
        :param params: instance of type "BuildMetabolicModelParams" ->
           structure: parameter "genome_id" of type "genome_id" (A string
           representing a Genome id.), parameter "genome_workspace" of type
           "workspace_name" (A string representing a workspace name.),
           parameter "media_id" of type "media_id" (A string representing a
           Media id.), parameter "media_workspace" of type "workspace_name"
           (A string representing a workspace name.), parameter
           "fbamodel_output_id" of type "fbamodel_id" (A string representing
           a FBAModel id.), parameter "workspace" of type "workspace_name" (A
           string representing a workspace name.), parameter "template_id" of
           type "template_id" (A string representing a NewModelTemplate id.),
           parameter "template_workspace" of type "workspace_name" (A string
           representing a workspace name.), parameter "coremodel" of type
           "bool" (A binary boolean), parameter "gapfill_model" of type
           "bool" (A binary boolean), parameter "thermodynamic_constraints"
           of type "bool" (A binary boolean), parameter
           "comprehensive_gapfill" of type "bool" (A binary boolean),
           parameter "custom_bound_list" of list of String, parameter
           "media_supplement_list" of list of type "compound_id" (A string
           representing a compound id.), parameter "expseries_id" of type
           "expseries_id" (A string representing an expression matrix id.),
           parameter "expseries_workspace" of type "workspace_name" (A string
           representing a workspace name.), parameter "expression_condition"
           of String, parameter "exp_threshold_percentile" of Double,
           parameter "exp_threshold_margin" of Double, parameter
           "activation_coefficient" of Double, parameter "omega" of Double,
           parameter "objective_fraction" of Double, parameter
           "minimum_target_flux" of Double, parameter "number_of_solutions"
           of Long
        :returns: instance of type "BuildMetabolicModelResults" -> structure:
           parameter "new_fbamodel_ref" of type "ws_fbamodel_id" (The
           workspace ID for a FBAModel data object. @id ws
           KBaseFBA.FBAModel), parameter "new_fba_ref" of type "ws_fba_id"
           (The workspace ID for a FBA data object. @id ws KBaseFBA.FBA),
           parameter "number_gapfilled_reactions" of Long, parameter
           "number_removed_biomass_compounds" of Long
        """
        return self._client.run_job('fba_tools.build_metabolic_model',
                                    [params], self._service_ver, context)

    def characterize_genome_metabolism_using_model(self, params, context=None):
        """
        Builds a model and characterizes an input genome using the model
        :param params: instance of type "CharacterizeGenomeUsingModelParams"
           -> structure: parameter "genome_id" of type "genome_id" (A string
           representing a Genome id.), parameter "genome_workspace" of type
           "workspace_name" (A string representing a workspace name.),
           parameter "fbamodel_output_id" of type "fbamodel_id" (A string
           representing a FBAModel id.), parameter "workspace" of type
           "workspace_name" (A string representing a workspace name.),
           parameter "template_id" of type "template_id" (A string
           representing a NewModelTemplate id.), parameter
           "template_workspace" of type "workspace_name" (A string
           representing a workspace name.), parameter
           "use_annotated_functions" of type "bool" (A binary boolean),
           parameter "merge_all_annotations" of type "bool" (A binary
           boolean), parameter "source_ontology_list" of list of String
        :returns: instance of type "CharacterizeGenomeUsingModelResults" ->
           structure: parameter "new_fbamodel_ref" of type "ws_fbamodel_id"
           (The workspace ID for a FBAModel data object. @id ws
           KBaseFBA.FBAModel), parameter "new_fba_ref" of type "ws_fba_id"
           (The workspace ID for a FBA data object. @id ws KBaseFBA.FBA)
        """
        return self._client.run_job('fba_tools.characterize_genome_metabolism_using_model',
                                    [params], self._service_ver, context)

    def run_model_characterization(self, params, context=None):
        """
        Builds a model and characterizes an input genome using the model
        :param params: instance of type "RunModelCharacterizationParams" ->
           structure: parameter "fbamodel_id" of type "fbamodel_id" (A string
           representing a FBAModel id.), parameter "fbamodel_workspace" of
           type "workspace_name" (A string representing a workspace name.),
           parameter "fbamodel_output_id" of type "fbamodel_id" (A string
           representing a FBAModel id.), parameter "workspace" of type
           "workspace_name" (A string representing a workspace name.)
        :returns: instance of type "RunModelCharacterizationResults" ->
           structure: parameter "new_fbamodel_ref" of type "ws_fbamodel_id"
           (The workspace ID for a FBAModel data object. @id ws
           KBaseFBA.FBAModel), parameter "new_fba_ref" of type "ws_fba_id"
           (The workspace ID for a FBA data object. @id ws KBaseFBA.FBA)
        """
        return self._client.run_job('fba_tools.run_model_characterization',
                                    [params], self._service_ver, context)

    def build_plant_metabolic_model(self, params, context=None):
        """
        Build a genome-scale metabolic model based on annotations in an input genome typed object
        :param params: instance of type "BuildPlantMetabolicModelParams" ->
           structure: parameter "genome_id" of type "genome_id" (A string
           representing a Genome id.), parameter "genome_workspace" of type
           "workspace_name" (A string representing a workspace name.),
           parameter "fbamodel_output_id" of type "fbamodel_id" (A string
           representing a FBAModel id.), parameter "workspace" of type
           "workspace_name" (A string representing a workspace name.),
           parameter "template_id" of type "template_id" (A string
           representing a NewModelTemplate id.), parameter
           "template_workspace" of type "workspace_name" (A string
           representing a workspace name.)
        :returns: instance of type "BuildPlantMetabolicModelResults" ->
           structure: parameter "new_fbamodel_ref" of type "ws_fbamodel_id"
           (The workspace ID for a FBAModel data object. @id ws
           KBaseFBA.FBAModel)
        """
        return self._client.run_job('fba_tools.build_plant_metabolic_model',
                                    [params], self._service_ver, context)

    def build_multiple_metabolic_models(self, params, context=None):
        """
        Build multiple genome-scale metabolic models based on annotations in an input genome typed object
        :param params: instance of type "BuildMultipleMetabolicModelsParams"
           -> structure: parameter "genome_ids" of list of type "genome_id"
           (A string representing a Genome id.), parameter "genome_text" of
           String, parameter "genome_workspace" of type "workspace_name" (A
           string representing a workspace name.), parameter "media_id" of
           type "media_id" (A string representing a Media id.), parameter
           "media_workspace" of type "workspace_name" (A string representing
           a workspace name.), parameter "fbamodel_output_id" of type
           "fbamodel_id" (A string representing a FBAModel id.), parameter
           "workspace" of type "workspace_name" (A string representing a
           workspace name.), parameter "template_id" of type "template_id" (A
           string representing a NewModelTemplate id.), parameter
           "template_workspace" of type "workspace_name" (A string
           representing a workspace name.), parameter "coremodel" of type
           "bool" (A binary boolean), parameter "gapfill_model" of type
           "bool" (A binary boolean), parameter "thermodynamic_constraints"
           of type "bool" (A binary boolean), parameter
           "comprehensive_gapfill" of type "bool" (A binary boolean),
           parameter "custom_bound_list" of list of String, parameter
           "media_supplement_list" of list of type "compound_id" (A string
           representing a compound id.), parameter "expseries_id" of type
           "expseries_id" (A string representing an expression matrix id.),
           parameter "expseries_workspace" of type "workspace_name" (A string
           representing a workspace name.), parameter "expression_condition"
           of String, parameter "exp_threshold_percentile" of Double,
           parameter "exp_threshold_margin" of Double, parameter
           "activation_coefficient" of Double, parameter "omega" of Double,
           parameter "objective_fraction" of Double, parameter
           "minimum_target_flux" of Double, parameter "number_of_solutions"
           of Long
        :returns: instance of type "BuildMultipleMetabolicModelsResults" ->
           structure: parameter "new_fbamodel_ref" of type "ws_fbamodel_id"
           (The workspace ID for a FBAModel data object. @id ws
           KBaseFBA.FBAModel), parameter "new_fba_ref" of type "ws_fba_id"
           (The workspace ID for a FBA data object. @id ws KBaseFBA.FBA)
        """
        return self._client.run_job('fba_tools.build_multiple_metabolic_models',
                                    [params], self._service_ver, context)

    def gapfill_metabolic_model(self, params, context=None):
        """
        Gapfills a metabolic model to induce flux in a specified reaction
        :param params: instance of type "GapfillMetabolicModelParams" ->
           structure: parameter "fbamodel_id" of type "fbamodel_id" (A string
           representing a FBAModel id.), parameter "fbamodel_workspace" of
           type "workspace_name" (A string representing a workspace name.),
           parameter "media_id" of type "media_id" (A string representing a
           Media id.), parameter "media_workspace" of type "workspace_name"
           (A string representing a workspace name.), parameter
           "target_reaction" of type "reaction_id" (A string representing a
           reaction id.), parameter "fbamodel_output_id" of type
           "fbamodel_id" (A string representing a FBAModel id.), parameter
           "workspace" of type "workspace_name" (A string representing a
           workspace name.), parameter "thermodynamic_constraints" of type
           "bool" (A binary boolean), parameter "comprehensive_gapfill" of
           type "bool" (A binary boolean), parameter "source_fbamodel_id" of
           type "fbamodel_id" (A string representing a FBAModel id.),
           parameter "source_fbamodel_workspace" of type "workspace_name" (A
           string representing a workspace name.), parameter
           "feature_ko_list" of list of type "feature_id" (A string
           representing a feature id.), parameter "reaction_ko_list" of list
           of type "reaction_id" (A string representing a reaction id.),
           parameter "custom_bound_list" of list of String, parameter
           "media_supplement_list" of list of type "compound_id" (A string
           representing a compound id.), parameter "expseries_id" of type
           "expseries_id" (A string representing an expression matrix id.),
           parameter "expseries_workspace" of type "workspace_name" (A string
           representing a workspace name.), parameter "expression_condition"
           of String, parameter "exp_threshold_percentile" of Double,
           parameter "exp_threshold_margin" of Double, parameter
           "activation_coefficient" of Double, parameter "omega" of Double,
           parameter "objective_fraction" of Double, parameter
           "minimum_target_flux" of Double, parameter "number_of_solutions"
           of Long
        :returns: instance of type "GapfillMetabolicModelResults" ->
           structure: parameter "new_fbamodel_ref" of type "ws_fbamodel_id"
           (The workspace ID for a FBAModel data object. @id ws
           KBaseFBA.FBAModel), parameter "new_fba_ref" of type "ws_fba_id"
           (The workspace ID for a FBA data object. @id ws KBaseFBA.FBA),
           parameter "number_gapfilled_reactions" of Long, parameter
           "number_removed_biomass_compounds" of Long
        """
        return self._client.run_job('fba_tools.gapfill_metabolic_model',
                                    [params], self._service_ver, context)

    def run_flux_balance_analysis(self, params, context=None):
        """
        Run flux balance analysis and return ID of FBA object with results
        :param params: instance of type "RunFluxBalanceAnalysisParams" ->
           structure: parameter "fbamodel_id" of type "fbamodel_id" (A string
           representing a FBAModel id.), parameter "fbamodel_workspace" of
           type "workspace_name" (A string representing a workspace name.),
           parameter "media_id" of type "media_id" (A string representing a
           Media id.), parameter "media_workspace" of type "workspace_name"
           (A string representing a workspace name.), parameter
           "target_reaction" of type "reaction_id" (A string representing a
           reaction id.), parameter "fba_output_id" of type "fba_id" (A
           string representing a FBA id.), parameter "workspace" of type
           "workspace_name" (A string representing a workspace name.),
           parameter "thermodynamic_constraints" of type "bool" (A binary
           boolean), parameter "fva" of type "bool" (A binary boolean),
           parameter "minimize_flux" of type "bool" (A binary boolean),
           parameter "simulate_ko" of type "bool" (A binary boolean),
           parameter "find_min_media" of type "bool" (A binary boolean),
           parameter "all_reversible" of type "bool" (A binary boolean),
           parameter "feature_ko_list" of list of type "feature_id" (A string
           representing a feature id.), parameter "reaction_ko_list" of list
           of type "reaction_id" (A string representing a reaction id.),
           parameter "custom_bound_list" of list of String, parameter
           "media_supplement_list" of list of type "compound_id" (A string
           representing a compound id.), parameter "expseries_id" of type
           "expseries_id" (A string representing an expression matrix id.),
           parameter "expseries_workspace" of type "workspace_name" (A string
           representing a workspace name.), parameter "expression_condition"
           of String, parameter "exp_threshold_percentile" of Double,
           parameter "exp_threshold_margin" of Double, parameter
           "activation_coefficient" of Double, parameter "omega" of Double,
           parameter "objective_fraction" of Double, parameter "max_c_uptake"
           of Double, parameter "max_n_uptake" of Double, parameter
           "max_p_uptake" of Double, parameter "max_s_uptake" of Double,
           parameter "max_o_uptake" of Double, parameter "default_max_uptake"
           of Double, parameter "notes" of String, parameter "massbalance" of
           String
        :returns: instance of type "RunFluxBalanceAnalysisResults" ->
           structure: parameter "new_fba_ref" of type "ws_fba_id" (The
           workspace ID for a FBA data object. @id ws KBaseFBA.FBA),
           parameter "objective" of Long, parameter "report_name" of String,
           parameter "report_ref" of type "ws_report_id" (The workspace ID
           for a Report object @id ws KBaseReport.Report)
        """
        return self._client.run_job('fba_tools.run_flux_balance_analysis',
                                    [params], self._service_ver, context)

    def compare_fba_solutions(self, params, context=None):
        """
        Compares multiple FBA solutions and saves comparison as a new object in the workspace
        :param params: instance of type "CompareFBASolutionsParams" ->
           structure: parameter "fba_id_list" of list of type "fba_id" (A
           string representing a FBA id.), parameter "fba_workspace" of type
           "workspace_name" (A string representing a workspace name.),
           parameter "fbacomparison_output_id" of type "fbacomparison_id" (A
           string representing a FBA comparison id.), parameter "workspace"
           of type "workspace_name" (A string representing a workspace name.)
        :returns: instance of type "CompareFBASolutionsResults" -> structure:
           parameter "new_fbacomparison_ref" of type "ws_fbacomparison_id"
           (The workspace ID for a FBA data object. @id ws KBaseFBA.FBA)
        """
        return self._client.run_job('fba_tools.compare_fba_solutions',
                                    [params], self._service_ver, context)

    def lookup_modelseed_ids(self, params, context=None):
        """
        Attempts to map peaks in the input metabolomics matrix to compounds in the ModelSEED database
        :param params: instance of type "LookupModelSEEDIDsParams" ->
           structure: parameter "workspace" of type "workspace_name" (A
           string representing a workspace name.), parameter
           "chemical_abundance_matrix_id" of type "metabolome_id" (A string
           representing a metabolome matrix id.), parameter
           "chemical_abundance_matrix_out_id" of type "metabolome_id" (A
           string representing a metabolome matrix id.)
        :returns: instance of type "LookupModelSEEDIDsResults" -> structure:
           parameter "report_name" of String, parameter "report_ref" of type
           "ws_report_id" (The workspace ID for a Report object @id ws
           KBaseReport.Report)
        """
        return self._client.run_job('fba_tools.lookup_modelseed_ids',
                                    [params], self._service_ver, context)

    def propagate_model_to_new_genome(self, params, context=None):
        """
        Translate the metabolic model of one organism to another, using a mapping of similar proteins between their genomes
        :param params: instance of type "PropagateModelToNewGenomeParams" ->
           structure: parameter "fbamodel_id" of type "fbamodel_id" (A string
           representing a FBAModel id.), parameter "fbamodel_workspace" of
           type "workspace_name" (A string representing a workspace name.),
           parameter "proteincomparison_id" of type "proteincomparison_id" (A
           string representing a protein comparison id.), parameter
           "proteincomparison_workspace" of type "workspace_name" (A string
           representing a workspace name.), parameter "fbamodel_output_id" of
           type "fbamodel_id" (A string representing a FBAModel id.),
           parameter "workspace" of type "workspace_name" (A string
           representing a workspace name.), parameter "keep_nogene_rxn" of
           type "bool" (A binary boolean), parameter "gapfill_model" of type
           "bool" (A binary boolean), parameter "media_id" of type "media_id"
           (A string representing a Media id.), parameter "media_workspace"
           of type "workspace_name" (A string representing a workspace
           name.), parameter "thermodynamic_constraints" of type "bool" (A
           binary boolean), parameter "comprehensive_gapfill" of type "bool"
           (A binary boolean), parameter "custom_bound_list" of list of
           String, parameter "media_supplement_list" of list of type
           "compound_id" (A string representing a compound id.), parameter
           "expseries_id" of type "expseries_id" (A string representing an
           expression matrix id.), parameter "expseries_workspace" of type
           "workspace_name" (A string representing a workspace name.),
           parameter "expression_condition" of String, parameter
           "translation_policy" of String, parameter
           "exp_threshold_percentile" of Double, parameter
           "exp_threshold_margin" of Double, parameter
           "activation_coefficient" of Double, parameter "omega" of Double,
           parameter "objective_fraction" of Double, parameter
           "minimum_target_flux" of Double, parameter "number_of_solutions"
           of Long
        :returns: instance of type "PropagateModelToNewGenomeResults" ->
           structure: parameter "new_fbamodel_ref" of type "ws_fbamodel_id"
           (The workspace ID for a FBAModel data object. @id ws
           KBaseFBA.FBAModel), parameter "new_fba_ref" of type "ws_fba_id"
           (The workspace ID for a FBA data object. @id ws KBaseFBA.FBA),
           parameter "number_gapfilled_reactions" of Long, parameter
           "number_removed_biomass_compounds" of Long
        """
        return self._client.run_job('fba_tools.propagate_model_to_new_genome',
                                    [params], self._service_ver, context)

    def simulate_growth_on_phenotype_data(self, params, context=None):
        """
        Use Flux Balance Analysis (FBA) to simulate multiple growth phenotypes.
        :param params: instance of type "SimulateGrowthOnPhenotypeDataParams"
           -> structure: parameter "fbamodel_id" of type "fbamodel_id" (A
           string representing a FBAModel id.), parameter
           "fbamodel_workspace" of type "workspace_name" (A string
           representing a workspace name.), parameter "phenotypeset_id" of
           type "phenotypeset_id" (A string representing a phenotype set
           id.), parameter "phenotypeset_workspace" of type "workspace_name"
           (A string representing a workspace name.), parameter
           "phenotypesim_output_id" of type "phenotypesim_id" (A string
           representing a phenotype simulation id.), parameter "workspace" of
           type "workspace_name" (A string representing a workspace name.),
           parameter "all_reversible" of type "bool" (A binary boolean),
           parameter "gapfill_phenotypes" of type "bool" (A binary boolean),
           parameter "fit_phenotype_data" of type "bool" (A binary boolean),
           parameter "save_fluxes" of type "bool" (A binary boolean),
           parameter "add_all_transporters" of type "bool" (A binary
           boolean), parameter "add_positive_transporters" of type "bool" (A
           binary boolean), parameter "target_reaction" of type "reaction_id"
           (A string representing a reaction id.), parameter
           "feature_ko_list" of list of type "feature_id" (A string
           representing a feature id.), parameter "reaction_ko_list" of list
           of type "reaction_id" (A string representing a reaction id.),
           parameter "custom_bound_list" of list of String, parameter
           "media_supplement_list" of list of type "compound_id" (A string
           representing a compound id.)
        :returns: instance of type "SimulateGrowthOnPhenotypeDataResults" ->
           structure: parameter "new_phenotypesim_ref" of type
           "ws_phenotypesim_id" (The workspace ID for a phenotype set
           simulation object. @id ws KBasePhenotypes.PhenotypeSimulationSet)
        """
        return self._client.run_job('fba_tools.simulate_growth_on_phenotype_data',
                                    [params], self._service_ver, context)

    def merge_metabolic_models_into_community_model(self, params, context=None):
        """
        Merge two or more metabolic models into a compartmentalized community model
        :param params: instance of type
           "MergeMetabolicModelsIntoCommunityModelParams" -> structure:
           parameter "fbamodel_id_list" of list of type "fbamodel_id" (A
           string representing a FBAModel id.), parameter
           "fbamodel_workspace" of type "workspace_name" (A string
           representing a workspace name.), parameter "fbamodel_output_id" of
           type "fbamodel_id" (A string representing a FBAModel id.),
           parameter "workspace" of type "workspace_name" (A string
           representing a workspace name.), parameter "mixed_bag_model" of
           type "bool" (A binary boolean)
        :returns: instance of type
           "MergeMetabolicModelsIntoCommunityModelResults" -> structure:
           parameter "new_fbamodel_ref" of type "ws_fbamodel_id" (The
           workspace ID for a FBAModel data object. @id ws KBaseFBA.FBAModel)
        """
        return self._client.run_job('fba_tools.merge_metabolic_models_into_community_model',
                                    [params], self._service_ver, context)

    def view_flux_network(self, params, context=None):
        """
        Merge two or more metabolic models into a compartmentalized community model
        :param params: instance of type "ViewFluxNetworkParams" -> structure:
           parameter "fba_id" of type "fba_id" (A string representing a FBA
           id.), parameter "fba_workspace" of type "workspace_name" (A string
           representing a workspace name.), parameter "workspace" of type
           "workspace_name" (A string representing a workspace name.)
        :returns: instance of type "ViewFluxNetworkResults" -> structure:
           parameter "new_report_ref" of type "ws_report_id" (The workspace
           ID for a Report object @id ws KBaseReport.Report)
        """
        return self._client.run_job('fba_tools.view_flux_network',
                                    [params], self._service_ver, context)

    def compare_flux_with_expression(self, params, context=None):
        """
        Merge two or more metabolic models into a compartmentalized community model
        :param params: instance of type "CompareFluxWithExpressionParams" ->
           structure: parameter "fba_id" of type "fba_id" (A string
           representing a FBA id.), parameter "fba_workspace" of type
           "workspace_name" (A string representing a workspace name.),
           parameter "expseries_id" of type "expseries_id" (A string
           representing an expression matrix id.), parameter
           "expseries_workspace" of type "workspace_name" (A string
           representing a workspace name.), parameter "expression_condition"
           of String, parameter "exp_threshold_percentile" of Double,
           parameter "estimate_threshold" of type "bool" (A binary boolean),
           parameter "maximize_agreement" of type "bool" (A binary boolean),
           parameter "fbapathwayanalysis_output_id" of type
           "fbapathwayanalysis_id" (A string representing a
           FBAPathwayAnalysis id.), parameter "workspace" of type
           "workspace_name" (A string representing a workspace name.)
        :returns: instance of type "CompareFluxWithExpressionResults" ->
           structure: parameter "new_fbapathwayanalysis_ref" of type
           "ws_fbapathwayanalysis_id" (The workspace ID for a FBA pathway
           analysis object @id ws KBaseFBA.FBAPathwayAnalysis)
        """
        return self._client.run_job('fba_tools.compare_flux_with_expression',
                                    [params], self._service_ver, context)

    def check_model_mass_balance(self, params, context=None):
        """
        Identifies reactions in the model that are not mass balanced
        :param params: instance of type "CheckModelMassBalanceParams" ->
           structure: parameter "fbamodel_id" of type "fbamodel_id" (A string
           representing a FBAModel id.), parameter "fbamodel_workspace" of
           type "workspace_name" (A string representing a workspace name.),
           parameter "workspace" of type "workspace_name" (A string
           representing a workspace name.)
        :returns: instance of type "CheckModelMassBalanceResults" ->
           structure: parameter "new_report_ref" of type "ws_report_id" (The
           workspace ID for a Report object @id ws KBaseReport.Report)
        """
        return self._client.run_job('fba_tools.check_model_mass_balance',
                                    [params], self._service_ver, context)

    def predict_auxotrophy(self, params, context=None):
        """
        Identifies reactions in the model that are not mass balanced
        :param params: instance of type "PredictAuxotrophyParams" ->
           structure: parameter "genome_ids" of list of type "genome_id" (A
           string representing a Genome id.), parameter "genome_workspace" of
           type "workspace_name" (A string representing a workspace name.),
           parameter "workspace" of type "workspace_name" (A string
           representing a workspace name.)
        :returns: instance of type "PredictAuxotrophyResults" -> structure:
           parameter "new_report_ref" of type "ws_report_id" (The workspace
           ID for a Report object @id ws KBaseReport.Report)
        """
        return self._client.run_job('fba_tools.predict_auxotrophy',
                                    [params], self._service_ver, context)

    def predict_metabolite_biosynthesis_pathway(self, params, context=None):
        """
        Identifies reactions in the model that are not mass balanced
        :param params: instance of type
           "PredictMetaboliteBiosynthesisPathwayInput" -> structure:
           parameter "fbamodel_id" of type "fbamodel_id" (A string
           representing a FBAModel id.), parameter "fbamodel_workspace" of
           type "workspace_name" (A string representing a workspace name.),
           parameter "media_id" of type "media_id" (A string representing a
           Media id.), parameter "media_workspace" of type "workspace_name"
           (A string representing a workspace name.), parameter
           "target_metabolite_list" of list of type "compound_id" (A string
           representing a compound id.), parameter "source_metabolite_list"
           of list of type "compound_id" (A string representing a compound
           id.), parameter "fba_output_id" of type "fba_id" (A string
           representing a FBA id.), parameter "workspace" of type
           "workspace_name" (A string representing a workspace name.),
           parameter "thermodynamic_constraints" of type "bool" (A binary
           boolean), parameter "feature_ko_list" of list of type "feature_id"
           (A string representing a feature id.), parameter
           "reaction_ko_list" of list of type "reaction_id" (A string
           representing a reaction id.), parameter "expseries_id" of type
           "expseries_id" (A string representing an expression matrix id.),
           parameter "expseries_workspace" of type "workspace_name" (A string
           representing a workspace name.), parameter "expression_condition"
           of String, parameter "exp_threshold_percentile" of Double,
           parameter "exp_threshold_margin" of Double, parameter
           "activation_coefficient" of Double, parameter "omega" of Double
        :returns: instance of type
           "PredictMetaboliteBiosynthesisPathwayResults" -> structure:
           parameter "report_name" of String, parameter "report_ref" of type
           "ws_report_id" (The workspace ID for a Report object @id ws
           KBaseReport.Report)
        """
        return self._client.run_job('fba_tools.predict_metabolite_biosynthesis_pathway',
                                    [params], self._service_ver, context)

    def build_metagenome_metabolic_model(self, params, context=None):
        """
        Build a genome-scale metabolic model based on annotations in an input genome typed object
        :param params: instance of type "BuildMetagenomeMetabolicModelParams"
           -> structure: parameter "input_ref" of String, parameter
           "input_workspace" of type "workspace_name" (A string representing
           a workspace name.), parameter "media_id" of type "media_id" (A
           string representing a Media id.), parameter "media_workspace" of
           type "workspace_name" (A string representing a workspace name.),
           parameter "fbamodel_output_id" of type "fbamodel_id" (A string
           representing a FBAModel id.), parameter "workspace" of type
           "workspace_name" (A string representing a workspace name.),
           parameter "gapfill_model" of type "bool" (A binary boolean),
           parameter "gff_file" of String
        :returns: instance of type "BuildMetabolicModelResults" -> structure:
           parameter "new_fbamodel_ref" of type "ws_fbamodel_id" (The
           workspace ID for a FBAModel data object. @id ws
           KBaseFBA.FBAModel), parameter "new_fba_ref" of type "ws_fba_id"
           (The workspace ID for a FBA data object. @id ws KBaseFBA.FBA),
           parameter "number_gapfilled_reactions" of Long, parameter
           "number_removed_biomass_compounds" of Long
        """
        return self._client.run_job('fba_tools.build_metagenome_metabolic_model',
                                    [params], self._service_ver, context)

    def build_model_from_roles(self, params, context=None):
        """
        Build a genome-scale metabolic model based on annotations in an input genome typed object
        :param params: instance of type "BuildModelFromRolesParams" ->
           structure: parameter "workspace" of type "workspace_name" (A
           string representing a workspace name.), parameter "roles" of type
           "ModelReconstructionInput" (Input to model reconstruction) ->
           structure: parameter "reaction_hash" of mapping from String to
           mapping from String to type "FunctionMappingData" -> structure:
           parameter "features" of mapping from String to type
           "FeatureMapping" -> structure: parameter "feature_ref" of String,
           parameter "probability" of Double, parameter "coverage" of Double,
           parameter "sources" of mapping from String to String, parameter
           "hit_count" of Long, parameter "non_gene_probability" of Double,
           parameter "non_gene_coverage" of Double, parameter "sources" of
           mapping from String to String, parameter "function_hash" of
           mapping from String to mapping from String to type
           "FunctionMappingData" -> structure: parameter "features" of
           mapping from String to type "FeatureMapping" -> structure:
           parameter "feature_ref" of String, parameter "probability" of
           Double, parameter "coverage" of Double, parameter "sources" of
           mapping from String to String, parameter "hit_count" of Long,
           parameter "non_gene_probability" of Double, parameter
           "non_gene_coverage" of Double, parameter "sources" of mapping from
           String to String, parameter "media_id" of type "media_id" (A
           string representing a Media id.), parameter "media_workspace" of
           type "workspace_name" (A string representing a workspace name.),
           parameter "fbamodel_output_id" of type "fbamodel_id" (A string
           representing a FBAModel id.), parameter "model_name" of String,
           parameter "template_id" of type "template_id" (A string
           representing a NewModelTemplate id.), parameter
           "template_workspace" of type "workspace_name" (A string
           representing a workspace name.), parameter "save_model" of type
           "bool" (A binary boolean)
        :returns: instance of type "BuildMetabolicModelResults" -> structure:
           parameter "new_fbamodel_ref" of type "ws_fbamodel_id" (The
           workspace ID for a FBAModel data object. @id ws
           KBaseFBA.FBAModel), parameter "new_fba_ref" of type "ws_fba_id"
           (The workspace ID for a FBA data object. @id ws KBaseFBA.FBA),
           parameter "number_gapfilled_reactions" of Long, parameter
           "number_removed_biomass_compounds" of Long
        """
        return self._client.run_job('fba_tools.build_model_from_roles',
                                    [params], self._service_ver, context)

    def fit_exometabolite_data(self, params, context=None):
        """
        Gapfills a metabolic model to fit input exometabolite data
        :param params: instance of type "FitExometaboliteDataParams" ->
           structure: parameter "fbamodel_id" of type "fbamodel_id" (A string
           representing a FBAModel id.), parameter "fbamodel_workspace" of
           type "workspace_name" (A string representing a workspace name.),
           parameter "source_fbamodel_id" of type "fbamodel_id" (A string
           representing a FBAModel id.), parameter
           "source_fbamodel_workspace" of type "workspace_name" (A string
           representing a workspace name.), parameter "media_id" of type
           "media_id" (A string representing a Media id.), parameter
           "media_workspace" of type "workspace_name" (A string representing
           a workspace name.), parameter "metabolome_id" of type
           "metabolome_id" (A string representing a metabolome matrix id.),
           parameter "metabolome_workspace" of type "workspace_name" (A
           string representing a workspace name.), parameter
           "metabolome_condition" of String, parameter "fbamodel_output_id"
           of type "fbamodel_id" (A string representing a FBAModel id.),
           parameter "workspace" of type "workspace_name" (A string
           representing a workspace name.), parameter "minimum_target_flux"
           of Double, parameter "omnidirectional" of type "bool" (A binary
           boolean), parameter "target_reaction" of type "reaction_id" (A
           string representing a reaction id.), parameter "feature_ko_list"
           of list of type "feature_id" (A string representing a feature
           id.), parameter "reaction_ko_list" of list of type "reaction_id"
           (A string representing a reaction id.), parameter
           "media_supplement_list" of list of type "compound_id" (A string
           representing a compound id.)
        :returns: instance of type "FitExometaboliteDataResults" ->
           structure: parameter "new_fbamodel_ref" of type "ws_fbamodel_id"
           (The workspace ID for a FBAModel data object. @id ws
           KBaseFBA.FBAModel), parameter "new_fba_ref" of type "ws_fba_id"
           (The workspace ID for a FBA data object. @id ws KBaseFBA.FBA),
           parameter "number_gapfilled_reactions" of Long
        """
        return self._client.run_job('fba_tools.fit_exometabolite_data',
                                    [params], self._service_ver, context)

    def compare_models(self, params, context=None):
        """
        Compare models
        :param params: instance of type "ModelComparisonParams"
           (ModelComparisonParams object: a list of models and optional
           pangenome and protein comparison; mc_name is the name for the new
           object. @optional protcomp_ref pangenome_ref) -> structure:
           parameter "workspace" of type "workspace_name" (A string
           representing a workspace name.), parameter "mc_name" of String,
           parameter "model_refs" of list of type "ws_fbamodel_id" (The
           workspace ID for a FBAModel data object. @id ws
           KBaseFBA.FBAModel), parameter "protcomp_ref" of type
           "ws_proteomecomparison_id" (Reference to a Proteome Comparison
           object in the workspace @id ws
           GenomeComparison.ProteomeComparison), parameter "pangenome_ref" of
           type "ws_pangenome_id" (Reference to a Pangenome object in the
           workspace @id ws KBaseGenomes.Pangenome)
        :returns: instance of type "ModelComparisonResult" -> structure:
           parameter "report_name" of String, parameter "report_ref" of type
           "ws_report_id" (The workspace ID for a Report object @id ws
           KBaseReport.Report), parameter "mc_ref" of String
        """
        return self._client.run_job('fba_tools.compare_models',
                                    [params], self._service_ver, context)

    def edit_metabolic_model(self, params, context=None):
        """
        Edit models
        :param params: instance of type "EditMetabolicModelParams"
           (EditMetabolicModelParams object: arguments for the edit model
           function) -> structure: parameter "workspace" of type
           "workspace_name" (A string representing a workspace name.),
           parameter "fbamodel_workspace" of type "workspace_name" (A string
           representing a workspace name.), parameter "fbamodel_id" of type
           "ws_fbamodel_id" (The workspace ID for a FBAModel data object. @id
           ws KBaseFBA.FBAModel), parameter "fbamodel_output_id" of type
           "ws_fbamodel_id" (The workspace ID for a FBAModel data object. @id
           ws KBaseFBA.FBAModel), parameter "compounds_to_add" of list of
           mapping from String to String, parameter "compounds_to_change" of
           list of mapping from String to String, parameter
           "biomasses_to_add" of list of mapping from String to String,
           parameter "biomass_compounds_to_change" of list of mapping from
           String to String, parameter "reactions_to_remove" of list of
           mapping from String to String, parameter "reactions_to_change" of
           list of mapping from String to String, parameter
           "reactions_to_add" of list of mapping from String to String,
           parameter "edit_compound_stoichiometry" of list of mapping from
           String to String
        :returns: instance of type "EditMetabolicModelResult" -> structure:
           parameter "report_name" of String, parameter "report_ref" of type
           "ws_report_id" (The workspace ID for a Report object @id ws
           KBaseReport.Report), parameter "new_fbamodel_ref" of type
           "ws_fbamodel_id" (The workspace ID for a FBAModel data object. @id
           ws KBaseFBA.FBAModel)
        """
        return self._client.run_job('fba_tools.edit_metabolic_model',
                                    [params], self._service_ver, context)

    def edit_media(self, params, context=None):
        """
        Edit models
        :param params: instance of type "EditMediaParams" (EditMediaParams
           object: arguments for the edit model function) -> structure:
           parameter "workspace" of type "workspace_name" (A string
           representing a workspace name.), parameter "media_id" of type
           "media_id" (A string representing a Media id.), parameter
           "media_workspace" of type "workspace_name" (A string representing
           a workspace name.), parameter "compounds_to_remove" of list of
           type "compound_id" (A string representing a compound id.),
           parameter "compounds_to_change" of list of tuple of size 4: type
           "compound_id" (A string representing a compound id.), parameter
           "concentration" of Double, parameter "min_flux" of Double,
           parameter "max_flux" of Double, parameter "compounds_to_add" of
           list of tuple of size 4: type "compound_id" (A string representing
           a compound id.), parameter "concentration" of Double, parameter
           "min_flux" of Double, parameter "max_flux" of Double, parameter
           "pH_data" of String, parameter "temperature" of Double, parameter
           "isDefined" of type "bool" (A binary boolean), parameter "type" of
           String, parameter "media_output_id" of type "media_id" (A string
           representing a Media id.)
        :returns: instance of type "EditMediaResult" -> structure: parameter
           "report_name" of String, parameter "report_ref" of type
           "ws_report_id" (The workspace ID for a Report object @id ws
           KBaseReport.Report), parameter "new_media_id" of type "media_id"
           (A string representing a Media id.)
        """
        return self._client.run_job('fba_tools.edit_media',
                                    [params], self._service_ver, context)

    def excel_file_to_model(self, p, context=None):
        """
        :param p: instance of type "ModelCreationParams" (compounds_file is
           not used for excel file creations) -> structure: parameter
           "model_file" of type "File" -> structure: parameter "path" of
           String, parameter "shock_id" of String, parameter "model_name" of
           String, parameter "workspace_name" of String, parameter "genome"
           of String, parameter "biomass" of list of String, parameter
           "compounds_file" of type "File" -> structure: parameter "path" of
           String, parameter "shock_id" of String
        :returns: instance of type "WorkspaceRef" -> structure: parameter
           "ref" of String
        """
        return self._client.run_job('fba_tools.excel_file_to_model',
                                    [p], self._service_ver, context)

    def sbml_file_to_model(self, p, context=None):
        """
        :param p: instance of type "ModelCreationParams" (compounds_file is
           not used for excel file creations) -> structure: parameter
           "model_file" of type "File" -> structure: parameter "path" of
           String, parameter "shock_id" of String, parameter "model_name" of
           String, parameter "workspace_name" of String, parameter "genome"
           of String, parameter "biomass" of list of String, parameter
           "compounds_file" of type "File" -> structure: parameter "path" of
           String, parameter "shock_id" of String
        :returns: instance of type "WorkspaceRef" -> structure: parameter
           "ref" of String
        """
        return self._client.run_job('fba_tools.sbml_file_to_model',
                                    [p], self._service_ver, context)

    def tsv_file_to_model(self, p, context=None):
        """
        :param p: instance of type "ModelCreationParams" (compounds_file is
           not used for excel file creations) -> structure: parameter
           "model_file" of type "File" -> structure: parameter "path" of
           String, parameter "shock_id" of String, parameter "model_name" of
           String, parameter "workspace_name" of String, parameter "genome"
           of String, parameter "biomass" of list of String, parameter
           "compounds_file" of type "File" -> structure: parameter "path" of
           String, parameter "shock_id" of String
        :returns: instance of type "WorkspaceRef" -> structure: parameter
           "ref" of String
        """
        return self._client.run_job('fba_tools.tsv_file_to_model',
                                    [p], self._service_ver, context)

    def model_to_excel_file(self, model, context=None):
        """
        :param model: instance of type "ModelObjectSelectionParams" ->
           structure: parameter "workspace_name" of String, parameter
           "model_name" of String, parameter "save_to_shock" of type
           "boolean" (A boolean - 0 for false, 1 for true. @range (0, 1)),
           parameter "fulldb" of type "bool" (A binary boolean)
        :returns: instance of type "File" -> structure: parameter "path" of
           String, parameter "shock_id" of String
        """
        return self._client.run_job('fba_tools.model_to_excel_file',
                                    [model], self._service_ver, context)

    def model_to_sbml_file(self, model, context=None):
        """
        :param model: instance of type "ModelObjectSelectionParams" ->
           structure: parameter "workspace_name" of String, parameter
           "model_name" of String, parameter "save_to_shock" of type
           "boolean" (A boolean - 0 for false, 1 for true. @range (0, 1)),
           parameter "fulldb" of type "bool" (A binary boolean)
        :returns: instance of type "File" -> structure: parameter "path" of
           String, parameter "shock_id" of String
        """
        return self._client.run_job('fba_tools.model_to_sbml_file',
                                    [model], self._service_ver, context)

    def model_to_tsv_file(self, model, context=None):
        """
        :param model: instance of type "ModelObjectSelectionParams" ->
           structure: parameter "workspace_name" of String, parameter
           "model_name" of String, parameter "save_to_shock" of type
           "boolean" (A boolean - 0 for false, 1 for true. @range (0, 1)),
           parameter "fulldb" of type "bool" (A binary boolean)
        :returns: instance of type "ModelTsvFiles" -> structure: parameter
           "compounds_file" of type "File" -> structure: parameter "path" of
           String, parameter "shock_id" of String, parameter "reactions_file"
           of type "File" -> structure: parameter "path" of String, parameter
           "shock_id" of String
        """
        return self._client.run_job('fba_tools.model_to_tsv_file',
                                    [model], self._service_ver, context)

    def export_model_as_excel_file(self, params, context=None):
        """
        :param params: instance of type "ExportParams" (input and output
           structure functions for standard downloaders) -> structure:
           parameter "input_ref" of String
        :returns: instance of type "ExportOutput" -> structure: parameter
           "shock_id" of String
        """
        return self._client.run_job('fba_tools.export_model_as_excel_file',
                                    [params], self._service_ver, context)

    def export_model_as_tsv_file(self, params, context=None):
        """
        :param params: instance of type "ExportParams" (input and output
           structure functions for standard downloaders) -> structure:
           parameter "input_ref" of String
        :returns: instance of type "ExportOutput" -> structure: parameter
           "shock_id" of String
        """
        return self._client.run_job('fba_tools.export_model_as_tsv_file',
                                    [params], self._service_ver, context)

    def export_model_as_sbml_file(self, params, context=None):
        """
        :param params: instance of type "ExportParams" (input and output
           structure functions for standard downloaders) -> structure:
           parameter "input_ref" of String
        :returns: instance of type "ExportOutput" -> structure: parameter
           "shock_id" of String
        """
        return self._client.run_job('fba_tools.export_model_as_sbml_file',
                                    [params], self._service_ver, context)

    def fba_to_excel_file(self, fba, context=None):
        """
        :param fba: instance of type "FBAObjectSelectionParams" (****** FBA
           Result Converters ******) -> structure: parameter "workspace_name"
           of String, parameter "fba_name" of String, parameter
           "save_to_shock" of type "boolean" (A boolean - 0 for false, 1 for
           true. @range (0, 1))
        :returns: instance of type "File" -> structure: parameter "path" of
           String, parameter "shock_id" of String
        """
        return self._client.run_job('fba_tools.fba_to_excel_file',
                                    [fba], self._service_ver, context)

    def fba_to_tsv_file(self, fba, context=None):
        """
        :param fba: instance of type "FBAObjectSelectionParams" (****** FBA
           Result Converters ******) -> structure: parameter "workspace_name"
           of String, parameter "fba_name" of String, parameter
           "save_to_shock" of type "boolean" (A boolean - 0 for false, 1 for
           true. @range (0, 1))
        :returns: instance of type "FBATsvFiles" -> structure: parameter
           "compounds_file" of type "File" -> structure: parameter "path" of
           String, parameter "shock_id" of String, parameter "reactions_file"
           of type "File" -> structure: parameter "path" of String, parameter
           "shock_id" of String
        """
        return self._client.run_job('fba_tools.fba_to_tsv_file',
                                    [fba], self._service_ver, context)

    def export_fba_as_excel_file(self, params, context=None):
        """
        :param params: instance of type "ExportParams" (input and output
           structure functions for standard downloaders) -> structure:
           parameter "input_ref" of String
        :returns: instance of type "ExportOutput" -> structure: parameter
           "shock_id" of String
        """
        return self._client.run_job('fba_tools.export_fba_as_excel_file',
                                    [params], self._service_ver, context)

    def export_fba_as_tsv_file(self, params, context=None):
        """
        :param params: instance of type "ExportParams" (input and output
           structure functions for standard downloaders) -> structure:
           parameter "input_ref" of String
        :returns: instance of type "ExportOutput" -> structure: parameter
           "shock_id" of String
        """
        return self._client.run_job('fba_tools.export_fba_as_tsv_file',
                                    [params], self._service_ver, context)

    def tsv_file_to_media(self, p, context=None):
        """
        :param p: instance of type "MediaCreationParams" (****** Media
           Converters *********) -> structure: parameter "media_file" of type
           "File" -> structure: parameter "path" of String, parameter
           "shock_id" of String, parameter "media_name" of String, parameter
           "workspace_name" of String
        :returns: instance of type "WorkspaceRef" -> structure: parameter
           "ref" of String
        """
        return self._client.run_job('fba_tools.tsv_file_to_media',
                                    [p], self._service_ver, context)

    def excel_file_to_media(self, p, context=None):
        """
        :param p: instance of type "MediaCreationParams" (****** Media
           Converters *********) -> structure: parameter "media_file" of type
           "File" -> structure: parameter "path" of String, parameter
           "shock_id" of String, parameter "media_name" of String, parameter
           "workspace_name" of String
        :returns: instance of type "WorkspaceRef" -> structure: parameter
           "ref" of String
        """
        return self._client.run_job('fba_tools.excel_file_to_media',
                                    [p], self._service_ver, context)

    def media_to_tsv_file(self, media, context=None):
        """
        :param media: instance of type "MediaObjectSelectionParams" ->
           structure: parameter "workspace_name" of String, parameter
           "media_name" of String, parameter "save_to_shock" of type
           "boolean" (A boolean - 0 for false, 1 for true. @range (0, 1))
        :returns: instance of type "File" -> structure: parameter "path" of
           String, parameter "shock_id" of String
        """
        return self._client.run_job('fba_tools.media_to_tsv_file',
                                    [media], self._service_ver, context)

    def media_to_excel_file(self, media, context=None):
        """
        :param media: instance of type "MediaObjectSelectionParams" ->
           structure: parameter "workspace_name" of String, parameter
           "media_name" of String, parameter "save_to_shock" of type
           "boolean" (A boolean - 0 for false, 1 for true. @range (0, 1))
        :returns: instance of type "File" -> structure: parameter "path" of
           String, parameter "shock_id" of String
        """
        return self._client.run_job('fba_tools.media_to_excel_file',
                                    [media], self._service_ver, context)

    def export_media_as_excel_file(self, params, context=None):
        """
        :param params: instance of type "ExportParams" (input and output
           structure functions for standard downloaders) -> structure:
           parameter "input_ref" of String
        :returns: instance of type "ExportOutput" -> structure: parameter
           "shock_id" of String
        """
        return self._client.run_job('fba_tools.export_media_as_excel_file',
                                    [params], self._service_ver, context)

    def export_media_as_tsv_file(self, params, context=None):
        """
        :param params: instance of type "ExportParams" (input and output
           structure functions for standard downloaders) -> structure:
           parameter "input_ref" of String
        :returns: instance of type "ExportOutput" -> structure: parameter
           "shock_id" of String
        """
        return self._client.run_job('fba_tools.export_media_as_tsv_file',
                                    [params], self._service_ver, context)

    def tsv_file_to_phenotype_set(self, p, context=None):
        """
        :param p: instance of type "PhenotypeSetCreationParams" (******
           Phenotype Data Converters *******) -> structure: parameter
           "phenotype_set_file" of type "File" -> structure: parameter "path"
           of String, parameter "shock_id" of String, parameter
           "phenotype_set_name" of String, parameter "workspace_name" of
           String, parameter "genome" of String
        :returns: instance of type "WorkspaceRef" -> structure: parameter
           "ref" of String
        """
        return self._client.run_job('fba_tools.tsv_file_to_phenotype_set',
                                    [p], self._service_ver, context)

    def phenotype_set_to_tsv_file(self, phenotype_set, context=None):
        """
        :param phenotype_set: instance of type
           "PhenotypeSetObjectSelectionParams" -> structure: parameter
           "workspace_name" of String, parameter "phenotype_set_name" of
           String, parameter "save_to_shock" of type "boolean" (A boolean - 0
           for false, 1 for true. @range (0, 1))
        :returns: instance of type "File" -> structure: parameter "path" of
           String, parameter "shock_id" of String
        """
        return self._client.run_job('fba_tools.phenotype_set_to_tsv_file',
                                    [phenotype_set], self._service_ver, context)

    def export_phenotype_set_as_tsv_file(self, params, context=None):
        """
        :param params: instance of type "ExportParams" (input and output
           structure functions for standard downloaders) -> structure:
           parameter "input_ref" of String
        :returns: instance of type "ExportOutput" -> structure: parameter
           "shock_id" of String
        """
        return self._client.run_job('fba_tools.export_phenotype_set_as_tsv_file',
                                    [params], self._service_ver, context)

    def phenotype_simulation_set_to_excel_file(self, pss, context=None):
        """
        :param pss: instance of type
           "PhenotypeSimulationSetObjectSelectionParams" -> structure:
           parameter "workspace_name" of String, parameter
           "phenotype_simulation_set_name" of String, parameter
           "save_to_shock" of type "boolean" (A boolean - 0 for false, 1 for
           true. @range (0, 1))
        :returns: instance of type "File" -> structure: parameter "path" of
           String, parameter "shock_id" of String
        """
        return self._client.run_job('fba_tools.phenotype_simulation_set_to_excel_file',
                                    [pss], self._service_ver, context)

    def phenotype_simulation_set_to_tsv_file(self, pss, context=None):
        """
        :param pss: instance of type
           "PhenotypeSimulationSetObjectSelectionParams" -> structure:
           parameter "workspace_name" of String, parameter
           "phenotype_simulation_set_name" of String, parameter
           "save_to_shock" of type "boolean" (A boolean - 0 for false, 1 for
           true. @range (0, 1))
        :returns: instance of type "File" -> structure: parameter "path" of
           String, parameter "shock_id" of String
        """
        return self._client.run_job('fba_tools.phenotype_simulation_set_to_tsv_file',
                                    [pss], self._service_ver, context)

    def export_phenotype_simulation_set_as_excel_file(self, params, context=None):
        """
        :param params: instance of type "ExportParams" (input and output
           structure functions for standard downloaders) -> structure:
           parameter "input_ref" of String
        :returns: instance of type "ExportOutput" -> structure: parameter
           "shock_id" of String
        """
        return self._client.run_job('fba_tools.export_phenotype_simulation_set_as_excel_file',
                                    [params], self._service_ver, context)

    def export_phenotype_simulation_set_as_tsv_file(self, params, context=None):
        """
        :param params: instance of type "ExportParams" (input and output
           structure functions for standard downloaders) -> structure:
           parameter "input_ref" of String
        :returns: instance of type "ExportOutput" -> structure: parameter
           "shock_id" of String
        """
        return self._client.run_job('fba_tools.export_phenotype_simulation_set_as_tsv_file',
                                    [params], self._service_ver, context)

    def bulk_export_objects(self, params, context=None):
        """
        :param params: instance of type "BulkExportObjectsParams" ->
           structure: parameter "refs" of list of String, parameter
           "all_models" of type "bool" (A binary boolean), parameter
           "all_fba" of type "bool" (A binary boolean), parameter "all_media"
           of type "bool" (A binary boolean), parameter "all_phenotypes" of
           type "bool" (A binary boolean), parameter "all_phenosims" of type
           "bool" (A binary boolean), parameter "model_format" of String,
           parameter "fba_format" of String, parameter "media_format" of
           String, parameter "phenotype_format" of String, parameter
           "phenosim_format" of String, parameter "workspace" of String,
           parameter "report_workspace" of String
        :returns: instance of type "BulkExportObjectsResult" -> structure:
           parameter "report_name" of String, parameter "report_ref" of type
           "ws_report_id" (The workspace ID for a Report object @id ws
           KBaseReport.Report), parameter "ref" of String
        """
        return self._client.run_job('fba_tools.bulk_export_objects',
                                    [params], self._service_ver, context)

    def run_fba_tools_tests(self, params, context=None):
        """
        :param params: instance of type "RunFbaToolsTestsParams" ->
           structure: parameter "test_metagenomes" of type "bool" (A binary
           boolean), parameter "workspace" of String
        :returns: instance of type "RunFbaToolsTestsResult" -> structure:
           parameter "report_name" of String, parameter "report_ref" of type
           "ws_report_id" (The workspace ID for a Report object @id ws
           KBaseReport.Report), parameter "ref" of String
        """
        return self._client.run_job('fba_tools.run_fba_tools_tests',
                                    [params], self._service_ver, context)

    def status(self, context=None):
        return self._client.run_job('fba_tools.status',
                                    [], self._service_ver, context)
